### DISCUSSION TITLE
Quarkus and Testcontainers - Gunnar Morling

### DISCUSSION THREAD

**Discussion by blog-comment-importer** [gunnarmorling/discussions.morling.dev#279] [Announcements]
https://www.morling.dev/blog/quarkus-and-testcontainers/

**Comment by blog-comment-importer** [comment]
_**Péter Veres** wrote at 2021-02-11T13:20:15Z (comment imported from Disqus):_

<p>Great article, thank you! It is much more convenient than the static initializer approach yet I could not find examples of it in the official documentations.</p>

**Reply by blog-comment-importer** [reply]
_**Gunnar Morling** wrote at 2021-02-11T14:07:58Z (comment imported from Disqus):_

<p>Thanks, Péter, happy to hear the post was useful to you. Indeed, not finding concise info elsewhere was the motivation for me to write it.</p>

**Comment by blog-comment-importer** [comment]
_**Siva** wrote at 2021-09-22T23:46:33Z (comment imported from Disqus):_

<p>In the recent version of Quarkus it is even simplified further. Just configure the following properties in <a href="http://application.properties" rel="nofollow noopener" title="application.properties">application.properties</a> and while running tests it will automatically spin up Postgresql using Testcontainers.</p>

```
%test.quarkus.datasource.jdbc.driver=org.testcontainers.jdbc.ContainerDatabaseDriver
%test.quarkus.datasource.jdbc.url=jdbc:tc:postgresql:latest:///dbname<br>%test.quarkus.hibernate-orm.dialect=org.hibernate.dialect.PostgreSQL9Dialect
```

**Comment by nilyin** [comment]
The important thing is that if you want your quarkus app to use the port (assigned dynamically) of postgres testcontainer you should NOT to set (neither general no-prefix or %test.* ) properties for username, password and url. All you need (considering settings in pom.xml for postgres testcontainer and etc) is single line in application.properties file: quarkus.datasource.db-kind=postgresql

you may specify different setting for $dev. profile. It doesn't affect testing. 

For example my working settings are given below:

quarkus.datasource.db-kind=postgresql
quarkus.hibernate-orm.database.generation=drop-and-create
quarkus.hibernate-validator.fail-fast=true

%dev.quarkus.datasource.reactive.url=postgresql:///DBName
%dev.quarkus.datasource.username=dev_userame
%dev.quarkus.datasource.password=dev_password

quarkus.datasource.devservices.enabled=true

**Comment by Achyutha7** [comment]
How do I override the dev profile configurations through containers. I am able to run the container in dev mode but it isn't overriding the already built variables that use the @ifbuildproperty annotation

**Comment by soft4rchitecture** [comment]
Hey guys.
At work we happen to have a single test resource, a couchbase db spin up using test container, using your technique.
We are also using 2 different test profiles (@TestProfile(MyProfile.class) is an example) sprinkled acrosss our test classes.
Junit is in charge of "randomly but deterministically" mixing test classes and running them in that order.
Every time the next test class is using a different QuarkusProfile, Quarkus is starting a new server,
spin down and up a new DB. Which takes 30 seconds. 

Now we could say we need to control those switches. 
You can alphanumerically sort test classes run by name since junit 5,8,x, but this only happens when running maven in the CLI. 
Then you put the classes using test container in a package per profile, and you're good to go... almost ... because ...
Most developers run tests using eclipse or IntelliJ.
Each of those IDE has its own Junit runner implementation which I fail to configure to execute test classes alphanumerically too ( what about you?)

So that means, at the will of the default junit runner strategy for mixing test classes, Quarkus test pseudo randomly causes unmanageable and time consuming IT runs.

Would you say that running test container with Quarkus is fine as long as:
1. you're using a single testProfile 
2. you're only running your tests in the CLI using maven and a properly configured unit-platform 

Otherwise this will cause a massive waste of time in tests as they grow, and they'll, causing the demise of TDD in the project in the end... 
Thx for your opinion and help :)
